src/test/kotlin/org/rust/ide/hints/RsGenericParameterInfoHandlerTest.kt

```diff
@@ -0,0 +1,268 @@
+    """, "T", "", 0)
+
+    fun `test nested generics 2`() = checkByText("""
+        struct S<T> { field: T, }
+        fn foo<P, K>(x: T) {}
+        fn main() { foo::<S<i32>, /*caret*/>(); }
+    """, "P, K", "", 1)
+
+    fun `test parametrized bounds`() = checkByText("""
+        pub trait Trait<T> {}
+        struct S<T> { x: T, }
+        fn foo<T: Trait<S<i32>>>(x: T) {}
+        fn main() { foo::</*caret*/>(); }
+    """, "T: Trait<S<i32>>", "", 0)
+
+    fun `test dotted method call`() = checkByText("""
+        struct S;
+        impl<T> S { fn foo<T>(&self, x: T) {} }
+        fn main() {
+        let s = S{};
+        s.foo::</*caret*/>(5); }
+    """, "T", "", 0)
+
+    fun `test explicit self method call`() = checkByText("""
+        struct S;
+        impl<T> S { fn foo<T>(&self, x: T) {} }
+        fn main() {
+        let s = S{};
+        S::foo::</*caret*/>(&s, 5); }
+    """, "T", "", 0)
+
+    fun `test trait method`() = checkByText("""
+        trait Trait {
+            fn in_trait<T>(&self, x: T) {}
+            fn another<T>(y: T) {}
+        }
+        struct S;
+        impl Trait for S {}
+        fn main() {
+            let s = S {};
+            s.in_trait::</*caret*/>();
+        }
+    """, "T", "", 0)
+
+    fun `test trait method 2`() = checkByText("""
+        trait Trait {
+            fn in_trait<T>(&self, x: T) {}
+            fn another<T>(y: T) {}
+        }
+        struct S;
+        impl Trait for S {}
+        fn main() {
+            let s = S {};
+            S::another::</*caret*/>();
+        }
+    """, "T", "", 0)
+
+    fun `test in fn declaration`() = checkByText("""
+        fn foo</*caret*/>(x: T) {}
+    """, "", "", -1)
+
+    fun `test in struct declaration`() = checkByText("""
+        struct S</*caret*/> { field: T, }
+    """, "", "", -1)
+
+    fun `test in enum declaration`() = checkByText("""
+        enum E</*caret*/> { Foo(T), }
+    """, "", "", -1)
+
+    fun `test in trait declaration`() = checkByText("""
+        pub trait Trait</*caret*/> {
+            fn foo(x: T);
+        }
+    """, "", "", -1)
+
+    fun `test after impl`() = checkByText("""
+        pub trait Trait<T> {}
+        struct S;
+        impl</*caret*/> Trait<T> for S { }
+    """, "", "", -1)
+
+    fun `test in impl block`() = checkByText("""
+        pub trait Trait<T> {}
+        struct S<K> {x: K}
+        impl<P> Trait</*caret*/> for S<P> { }
+    """, "T", "", 0)
+
+    private fun checkByText(@Language("Rust") code: String, hint: String, where: String, index: Int) {
+        myFixture.configureByText("main.rs", replaceCaretMarker(code))
+        val handler = RsGenericParameterInfoHandler()
+        val createContext = MockCreateParameterInfoContext(myFixture.editor, myFixture.file)
+
+        // Check hint
+        val elt = handler.findElementForParameterInfo(createContext)
+        if (hint.isNotEmpty()) {
+            elt ?: throw AssertionFailedError("Hint not found")
```

Comment:
You can replace `throw AssertionFailedError(...)` to Kotlin's function `error(...)`